File: tests/e2e/pages/login_page.py
from playwright.sync_api import Page, Locator, expect
from .base_page import BasePage


class LoginPage(BasePage):
    """POM for TestFlow's auth modal (there is no dedicated /login or /signup route —
    sign-in is a modal triggered by the 'Sign in' button on any page).

    HTML landmarks (see showAuthModal/renderAuthForm in static/app.js):
      [data-testid=signin-btn]     — opens the modal
      [data-testid=auth-username]  — username input
      [data-testid=auth-password]  — password input
      [data-testid=auth-submit-btn]— submit button
      #auth-error                  — validation/auth failure message
    """

    def __init__(self, page: Page, base_url: str):
        super().__init__(page, base_url)

    # ── Locators ──────────────────────────────────────────────────────────────

    @property
    def heading(self) -> Locator:
        return self.page.locator("#auth-form-container h1", has_text="TestFlow")

    @property
    def sign_in_heading(self) -> Locator:
        return self.page.locator("#auth-form-container h2", has_text="Sign in")

    @property
    def username_input(self) -> Locator:
        return self.page.get_by_test_id("auth-username")

    @property
    def password_input(self) -> Locator:
        return self.page.get_by_test_id("auth-password")

    @property
    def submit_btn(self) -> Locator:
        return self.page.get_by_test_id("auth-submit-btn")

    @property
    def auth_error(self) -> Locator:
        return self.page.locator("#auth-error")

    @property
    def contact_admin_text(self) -> Locator:
        return self.page.locator("text=Contact your admin to get an account.")

    # ── Actions ───────────────────────────────────────────────────────────────

    def open(self):
        """Load the app (as a guest) and open the sign-in modal."""
        self.log.step("Open sign-in modal")
        self.goto()
        self.page.get_by_test_id("signin-btn").click()
        self.log.assert_("auth modal visible")
        expect(self.submit_btn).to_be_visible()

    def login(self, username: str, password: str):
        """Open the sign-in modal and submit credentials (does not assume success)."""
        self.log.step(f"Attempt login as '{username}'")
        self.open()
        self.log.action("fill", "username field", username)
        self.username_input.fill(username)
        self.log.action("fill", "password field", "***")
        self.password_input.fill(password)
        self.log.action("click", "Sign in submit button")
        self.submit_btn.click()

    # ── Assertions ────────────────────────────────────────────────────────────

    def expect_modal_loaded(self):
        self.log.assert_("auth modal fully rendered")
        expect(self.heading).to_be_visible()
        expect(self.sign_in_heading).to_be_visible()
        expect(self.username_input).to_be_visible()
        expect(self.password_input).to_be_visible()
        expect(self.submit_btn).to_be_visible()

    def expect_contact_admin_text_visible(self):
        self.log.assert_("'contact your admin' message visible")
        expect(self.contact_admin_text).to_be_visible()

    def expect_login_error(self, text: str):
        self.log.assert_("auth error visible", text)
        expect(self.auth_error).to_be_visible()
        expect(self.auth_error).to_contain_text(text)

    def expect_logged_in(self):
        self.log.assert_("logged in — logout button visible")
        expect(self.page.get_by_test_id("logout-btn")).to_be_visible()

File: e2e/pages/login.page.ts
import { Page, Locator, expect } from '@playwright/test';
import { BasePage } from './base.page';
import { log } from '../logger';

/**
 * POM for TestFlow's auth modal (there is no dedicated /login or /signup route —
 * sign-in is a modal triggered by the 'Sign in' button on any page).
 *
 * HTML landmarks (see showAuthModal/renderAuthForm in static/app.js):
 *   [data-testid=signin-btn]      — opens the modal
 *   [data-testid=auth-username]   — username input
 *   [data-testid=auth-password]   — password input
 *   [data-testid=auth-submit-btn] — submit button
 *   #auth-error                   — validation/auth failure message
 */
export class LoginPage extends BasePage {
  readonly heading:         Locator;
  readonly signInHeading:   Locator;
  readonly usernameInput:   Locator;
  readonly passwordInput:   Locator;
  readonly submitBtn:       Locator;
  readonly authError:       Locator;
  readonly contactAdminText: Locator;

  constructor(page: Page) {
    super(page);
    this.heading          = page.locator('#auth-form-container h1', { hasText: 'TestFlow' });
    this.signInHeading    = page.locator('#auth-form-container h2', { hasText: 'Sign in' });
    this.usernameInput    = page.getByTestId('auth-username');
    this.passwordInput    = page.getByTestId('auth-password');
    this.submitBtn        = page.getByTestId('auth-submit-btn');
    this.authError        = page.locator('#auth-error');
    this.contactAdminText = page.getByText('Contact your admin to get an account.');
  }

  /** Load the app as a guest and open the sign-in modal. */
  async open(): Promise<void> {
    log.step('Open sign-in modal');
    await this.navigate('/');
    await this.waitForNetworkIdle();
    await this.page.getByTestId('signin-btn').click();
    await expect(this.submitBtn).toBeVisible();
    log.assert('auth modal visible');
  }

  /** Open the sign-in modal and submit credentials (does not assume success). */
  async login(username: string, password: string): Promise<void> {
    log.step(`Attempt login as '${username}'`);
    await this.open();
    log.action('fill', 'username field', username);
    await this.usernameInput.fill(username);
    log.action('fill', 'password field', '***');
    await this.passwordInput.fill(password);
    log.action('click', 'Sign in submit button');
    await this.submitBtn.click();
  }

  async expectModalLoaded(): Promise<void> {
    log.assert('auth modal fully rendered');
    await expect(this.heading).toBeVisible();
    await expect(this.signInHeading).toBeVisible();
    await expect(this.usernameInput).toBeVisible();
    await expect(this.passwordInput).toBeVisible();
    await expect(this.submitBtn).toBeVisible();
  }

  async expectContactAdminTextVisible(): Promise<void> {
    log.assert("'contact your admin' message visible");
    await expect(this.contactAdminText).toBeVisible();
  }

  async expectLoginError(text: string): Promise<void> {
    log.assert('auth error visible', text);
    await expect(this.authError).toBeVisible();
    await expect(this.authError).toContainText(text);
  }

  async expectLoggedIn(): Promise<void> {
    log.assert('logged in — logout button visible');
    await expect(this.page.getByTestId('logout-btn')).toBeVisible();
  }
}
